### PR TITLE
 stats: implement Count-Min Sketch

### PR DESCRIPTION
**Pull request by lamxTyler**
Count-Min Sketch is used to estimate point queries. It consists of `d` counter array of size `w`. Every item will be mapped into one position in each of the array.
PTAL @coocood @hanfei1991 @winoros

### CODE REVIEW

**Comment by shenli** [comment]
Any reference for the Count-Min Sketch algorithm?

**Comment by hanfei1991** [comment]
@shenli A design document has been shared with tidb team by email.

File: statistics/cmsketch.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	h1, h2 := murmur3.Sum128(bytes)
+	vals := make([]uint32, c.depth)
+	min := uint32(math.MaxUint32)
+	for i := range c.table {
+		j := (h1 + h2*uint64(i)) % uint64(c.width)
+		if min > c.table[i][j] {
+			min = c.table[i][j]
+		}
+		noise := (c.count - uint64(c.table[i][j])) / (uint64(c.width) - 1)
+		if uint64(c.table[i][j]) < noise {
```
Comment (jackysp) [inline statistics/cmsketch.go]: Could this happen?

Comment (lamxTyler) [reply statistics/cmsketch.go]: Yes, it occurs quite a few.

Comment (jackysp) [reply statistics/cmsketch.go]: ok

File: statistics/cmsketch.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+func newCMSketch(d, w int32) *CMSketch {
+	tbl := make([][]uint32, d)
+	for i := range tbl {
+		tbl[i] = make([]uint32, w)
+	}
+	return &CMSketch{depth: d, width: w, table: tbl}
+}
+
+func (c *CMSketch) insert(val types.Datum) error {
```
Comment (zz-jason) [inline statistics/cmsketch.go]: how about use `*types.Datum` to avoid `runtime.duffcopy` ?

File: statistics/cmsketch.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	return nil
+}
+
+type uint32arr []uint32
+
+func (a uint32arr) Len() int           { return len(a) }
+func (a uint32arr) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a uint32arr) Less(i, j int) bool { return a[i] < a[j] }
+
+func (c *CMSketch) query(val types.Datum) (uint32, error) {
```
Comment (zz-jason) [inline statistics/cmsketch.go]: ditto

File: statistics/cmsketch.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	c.count++
+	h1, h2 := murmur3.Sum128(bytes)
+	for i := range c.table {
+		j := (h1 + h2*uint64(i)) % uint64(c.width)
+		c.table[i][j]++
+	}
+	return nil
+}
+
+type uint32arr []uint32
```
Comment (zz-jason) [inline statistics/cmsketch.go]: does it have to be `uint32` ? if we can use `int`, we can sort this int slice by `sort.Ints()`

File: statistics/cmsketch.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+func (c *CMSketch) query(val types.Datum) (uint32, error) {
+	bytes, err := codec.EncodeValue(nil, val)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	h1, h2 := murmur3.Sum128(bytes)
+	vals := make([]uint32, c.depth)
+	min := uint32(math.MaxUint32)
+	for i := range c.table {
+		j := (h1 + h2*uint64(i)) % uint64(c.width)
```
Comment (zz-jason) [inline statistics/cmsketch.go]: how about extracting `(h1 + h2*uint64(i)) % uint64(c.width)` as a function ?

File: statistics/cmsketch.go (hunk 4 above)
Comment (lamxTyler) [reply statistics/cmsketch.go]: Yes, `uint32` could represent more non-negative values than  `int`.

File: statistics/cmsketch.go (hunk 5 above)
Comment (lamxTyler) [reply statistics/cmsketch.go]: Is it too simple to be a function?

**Comment by wangqiim** [comment]
Could you please share the design document? In tinysql, CMS has been used as course project. @hanfei1991 @lamxTyler

**Pull request merged**